### PR TITLE
Update kal selection

### PR DESCRIPTION
**Pull request by MichaelGrosio**

### CODE REVIEW

File: projects/kalidea/kaligraphi/package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "@angular/animations": "^7.0.0",
     "@angular/forms": "^7.0.0",
     "rxjs": "~6.3.3",
+    "lodash": "latest",
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/package.json]: sure about that ?

File: projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-list.component.ts [hunk 2]
```diff
@@ -298,10 +330,14 @@ export class KalListComponent<T extends { id: string }> implements CollectionVie
   /**
    * Reset the selected item
    */
-  reset() {
-    this._selection = new KalListSelection<T>();
+  clear() {
+    // this._selection = new KalListSelection<T>();
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-list.component.ts]: comment

File: projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-list.component.ts [hunk 3]
```diff
@@ -314,11 +350,16 @@ export class KalListComponent<T extends { id: string }> implements CollectionVie
...
 
-  toggleItem(item: T) {
-    if (this._selection.indexOf(item) === -1) {
-      this._selection.add(item);
-    } else {
-      this._selection.remove(item);
+  /**
+   * Is the item highlighted
+   */
+  isHighlighted(item) {
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-list.component.ts]: type returns

File: projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-selection.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
+import { SelectionModel } from '@angular/cdk/collections';
+import { cloneDeep, isNil } from 'lodash';
+
+export class KalSelection<T extends { id?: string }> {
+
+  count ? = 0; // Total of elements
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-selection.ts]: Is the editor able to define theses properties with its comment ?
Maybe you should set comments with /** */

File: projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-selection.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  added?: T[] = []; // List of added elements
+
+  removed?: T[] = []; // List of removed elements
+
+  all ? = false; // Check if all items are selected
+
+  multiple ? = false;
+}
+
+class SubSelectionModel<T extends { id?: string }> extends SelectionModel<T> {
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-selection.ts]: This class is only used in Kaligraphi ?

File: projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-selection.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    this.removed = new SubSelectionModel<T>(multiple, [], true);
+    this.removed.select(...(removed as T[]));
+    this.all = all;
+    this.count = count;
+  }
+
+  /**
+   * get number of selected items
+   */
+  get total() {
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-selection.ts]: type returns

File: projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-selection.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    return {
+      added: this.added.selected,
+      removed: this.removed.selected,
+      all: this.all,
+      count: this.count,
+      total: this.total
+    };
+  }
+
+  changes() {
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-selection.ts]: empty function

**Review by jsp-up — CHANGES_REQUESTED** [review]

**Review by ydomenjoud — DISMISSED** [review]

File: projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-list.component.ts [hunk 8]
```diff
@@ -362,18 +383,21 @@ export class KalListComponent<T extends { id: string }> implements CollectionVie
       this.subscription = (this.dataSource as DataSource<T>).connect(this).subscribe(
         (items: T[]) => {
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-list.component.ts]: factorize subscribe

File: projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-selection.ts (hunk 5 above)
Comment (MichaelGrosio) [reply projects/kalidea/kaligraphi/src/lib/molecules/kal-list/kal-selection.ts]: Yes

**Review by jsp-up — DISMISSED** [review]

**Review by jsp-up — APPROVED** [review]

**Pull request merged**